docs/development/test-guidelines.adoc

```diff
@@ -0,0 +1,209 @@
+
+=== Blackbox vs. Whitebox tests
+
+In a blackbox test, you test a package as if you **only have access to the
+exported symbols**. In our <<lib-dot-go-source>> example above we have two exported
+symbols: `Say` and `Shout`. The `toUpper` function is an unexported symbol due
+to its lowercase name. 
+
+A whitebox test has access to all symbols, and not only to the expored ones.
+
+In principle, all symbols in a file can be *seen* by all files in the same package.
+As a consequence, to have a blackbox test, we must write it in a different package
+than `bar`. But in order for `go test` to pick up a test we need to place it in the
+same directory as the <<lib-dot-go-source>> file.
+
+In Go your test filenames must end with `_test.go`. In order to distinguish black-
+and whitebox test files, a recommendation is to use the suffixes `_blackbox_test.go`
+and `_whitebox_test.go`.
+
+[[lib-blackbox-test-dot-go-source]]
+[source,go]
+.$GOPATH/src/foo/bar/lib_blackbox_test.go
+----
+package bar_test // <1>
+
+import (
+	. "foo/bar" // <2>
+	"testing"
+)
+
+func TestSay(t *testing.T) { // <3>
+	t.Parallel() // <4>
+	if Say("heLLo") != "heLLo" { 
+		t.Errorf("Wrong say result")
+	}
+}
+
+func TestShout(t *testing.T) { // <5>
+	t.Parallel() // <6>
+	if Shout("heLLo") != "HELLO!!!" {
+		t.Errorf("Wrong shout result")
+	}
+}
+----
+<1> Notice that Go **allows** the package name `bar_test` to exist in the `bar` folder.
+<2> One of the very few circumstances in which a dot-import is okay in Go. Here it simply
+    allows for easier writing.
+<3> Testing the exported `Say` function.
+<4> Marking the test for the `Say` function test as: "executable in parallel".
+<5> Testing the exported `Shout` function.
+<6> Marking the test for the `Shout` function test as: "executable in parallel".
+
+A few things to notice in <<lib-blackbox-test-dot-go-source>> is the package name: `bar_test`.
+This is an exception to the Go compiler because normally only package names with the same name are
+allowed in one folder. Go doesn't complain if you add `_test` to the name though.
+A nice side-effect of this exception is that it allows us to better determine the coverage of
+a package. If the test functions needed to live inside of a different directory, we would
+not have been able to collect coverage information because currently Go can only collect
+this for a single package or directory.
+
+NOTE: Have you noticed the `t.Parallel()` calls? This is a hint for `go test` so that it knows
+that the test `t` can be executed in parallel. Normally `go test` executes tests sequentially,
+but it also tries to executes as many tests in parallel as there are CPUs and as there are tests
+that have been marked to be executable in parallel. In this example, none of our tests
+has any side effects and so they can be executed in parallel.
+The number of parallel executable tests can be overwritten by the `-parallel` switch.
+
```

Comment:
Since tests on this level should be independent, can we switch parallel execution globally? I.e. parallel would be default and we would need to set serial execution explicitly in special cases.